Sources/NIO/CircularBuffer.swift

```diff
@@ -79,6 +85,77 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
         }
     }
 
+    /// Replaces the specified subrange of elements with the given collection.
+    ///
+    /// - Parameter subrange:
+    /// The subrange of the collection to replace. The bounds of the range must be valid indices of the collection.
+    ///
+    /// - Parameter newElements:
+    /// The new elements to add to the collection.
+    ///
+    /// *O(n)* where _n_ is the length of the new elements collection if the subrange equals to _n_
+    ///
+    /// *O(m)* where _m_ is the combined length of the collection and _newElements_
+    public mutating func replaceSubrange<C>(_ subrange: Range<Index>, with newElements: C) where C : Collection, E == C.Element {
+        precondition(subrange.lowerBound >= self.startIndex && subrange.upperBound <= self.endIndex, "Subrange out of bounds")
+
+        if subrange.count == newElements.count {
+            for (index, element) in zip(stride(from: subrange.lowerBound, to: subrange.upperBound, by: 1), newElements) {
+                self.buffer[self.bufferIndex(ofIndex: index)] = element
+            }
+        } else if subrange.count == self.count && newElements.isEmpty {
+            self.removeSubrange(subrange)
+        } else {
+            var newBuffer: ContiguousArray<E?> = []
+            #if swift(>=4.2)
+            let capacityDelta = (newElements.count - subrange.count)
+            #else
+            let capacityDelta = (Int(newElements.count) - subrange.count)
+            #endif
+            let newCapacity = Int(UInt32(self.buffer.count + capacityDelta).nextPowerOf2())
+            newBuffer.reserveCapacity(newCapacity)
+
+            newBuffer.append(contentsOf: self[0..<subrange.lowerBound].map { $0 })
+            newBuffer.append(contentsOf: newElements.map { $0 })
+            newBuffer.append(contentsOf: self[subrange.upperBound..<self.endIndex].map { $0 })
```

Comment:
`append(contentsOf:)` takes `Sequence`, so there should be no need for the call to `map`.

Reply:
That was my first idea but the compiler doesn't let me and complains that the `contentsOf:` label is not needed 😕; which obviously is wrong because then I tries to append the sequence as if it was an array of sequences

Reply:
<img width="1245" alt="circularbuffer swift 2018-06-06 19-12-08" src="https://user-images.githubusercontent.com/2642850/41054051-961cbf0e-69bd-11e8-9713-3639add03b76.png">

Reply:
It makes sense though… because I'm slicing `CircularBuffer` and then try to append to `ContiguosArray` so I end up trying to add `Slice<CircularBuffer<E>>` to `ContiguosArray<E>`

Reply:
Well that's annoying. I just filed [SR-7921](https://bugs.swift.org/browse/SR-7921) to resolve the fact that this fixes the issue.

In the meantime, please update this with a comment indicating that this behaviour is silly.